File: backend/src/schemas/criminal.py
from typing import Optional, List
from datetime import date, datetime
from uuid import UUID
from pydantic import BaseModel
from src.domain.models.criminal import ThreatLevel, LegalStatus
from src.schemas.face_quality import FaceQualityResponse
from src.schemas.identity_template import IdentityTemplateResponse
from src.schemas.review_case import DuplicateReviewSummary

class CriminalBaseSchema(BaseModel):
    nic: Optional[str] = None
    first_name: str
    last_name: str
    aliases: Optional[str] = None
    dob: Optional[date] = None
    gender: str
    blood_type: Optional[str] = None
    last_known_address: Optional[str] = None
    status: LegalStatus = LegalStatus.WANTED
    threat_level: ThreatLevel = ThreatLevel.LOW
    physical_description: Optional[str] = None

class CriminalCreate(CriminalBaseSchema):
    pass

class CriminalUpdate(BaseModel):
    nic: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    aliases: Optional[str] = None
    dob: Optional[date] = None
    gender: Optional[str] = None
    blood_type: Optional[str] = None
    last_known_address: Optional[str] = None
    status: Optional[LegalStatus] = None
    threat_level: Optional[ThreatLevel] = None
    physical_description: Optional[str] = None

class CriminalResponse(CriminalBaseSchema):
    id: UUID
    primary_face_image_url: Optional[str] = None
    
    class Config:
        from_attributes = True

class CriminalListResponse(BaseModel):
    items: List[CriminalResponse]
    total: int
    page: int
    pages: int


class CriminalFaceResponse(BaseModel):
    id: UUID
    criminal_id: UUID
    image_url: str
    is_primary: bool
    embedding_version: str
    created_at: datetime
    box: tuple[int, int, int, int]
    exclude_from_template: bool = False
    operator_review_status: str = "normal"
    operator_review_notes: str | None = None
    template_role: str
    template_distance: float | None = None
    quality: FaceQualityResponse
    duplicate_review: DuplicateReviewSummary | None = None


class CriminalFaceReviewActionResponse(BaseModel):
    status: str
    message: str
    promoted_face_id: UUID | None = None
    face: CriminalFaceResponse


class CriminalTemplateRebuildResponse(BaseModel):
    status: str
    message: str
    criminal_id: UUID
    template: IdentityTemplateResponse | None = None

File: frontend/src/types/criminal.ts
import type { DuplicateReviewSummary } from '@/types/review';

export type FaceTemplateRole = 'primary' | 'support' | 'archived' | 'outlier' | (string & {});

export interface Criminal {
    id: string;
    first_name: string;
    last_name: string;
    aliases?: string;
    nic?: string;
    dob?: string;
    gender: string;
    blood_type?: string;
    status?: 'wanted' | 'in_custody' | 'released' | 'deceased' | 'cleared';
    threat_level?: 'low' | 'medium' | 'high' | 'critical';
    last_known_address?: string;
    physical_description?: string;
    primary_face_image_url?: string | null;
}

export interface CriminalFace {
    id: string;
    criminal_id: string;
    image_url: string;
    is_primary: boolean;
    embedding_version: string;
    created_at: string;
    box: [number, number, number, number];
    exclude_from_template: boolean;
    operator_review_status: string;
    operator_review_notes?: string | null;
    template_role: FaceTemplateRole;
    template_distance?: number | null;
    quality: FaceQuality;
    duplicate_review?: DuplicateReviewSummary | null;
}

export interface FaceQuality {
    status: 'accepted' | 'accepted_with_warnings' | 'rejected';
    quality_score: number;
    blur_score: number;
    brightness_score: number;
    face_area_ratio: number;
    pose_score: number;
    occlusion_score: number;
    warnings: string[];
}

export interface FaceQualityPreview {
    status: 'accepted' | 'accepted_with_warnings' | 'rejected';
    detected_face_count: number;
    decision_reason: string;
    message: string;
    box?: [number, number, number, number] | null;
    quality?: FaceQuality | null;
}

export interface CriminalFaceReviewActionResponse {
    status: string;
    message: string;
    promoted_face_id?: string | null;
    face: CriminalFace;
}

export interface CriminalTemplateRebuild {
    id: string;
    criminal_id: string;
    template_version: string;
    embedding_version: string;
    primary_face_id?: string | null;
    included_face_ids: string[];
    support_face_ids: string[];
    archived_face_ids: string[];
    outlier_face_ids: string[];
    active_face_count: number;
    support_face_count: number;
    archived_face_count: number;
    outlier_face_count: number;
    updated_at: string;
}

export interface CriminalTemplateRebuildResponse {
    status: string;
    message: string;
    criminal_id: string;
    template?: CriminalTemplateRebuild | null;
}

export interface CriminalFormData {
    first_name: string;
    last_name: string;
    aliases?: string;
    nic?: string;
    dob?: string;
    gender: string;
    blood_type?: string;
    status?: string;
    threat_level?: string;
    last_known_address?: string;
    physical_description?: string;
    faceFiles?: File[];
    primaryFaceIndex?: number;
}
